tests/common/dualtor/tunnel_traffic_utils.py

```diff
@@ -0,0 +1,131 @@
+            exp_pkt.set_do_not_care_scapy(IP, "ttl")
+            exp_pkt.set_do_not_care_scapy(IP, "proto")
+            exp_pkt.set_do_not_care_scapy(IP, "chksum")
+            exp_pkt.set_ignore_extra_bytes()
+            return exp_pkt
+
+        @staticmethod
+        def _dump_show_str(packet):
+            """Dump packet show output to string."""
+            _stdout, sys.stdout = sys.stdout, BytesIO()
+            try:
+                packet.show()
+                return sys.stdout.getvalue()
+            finally:
+                sys.stdout = _stdout
+
+        @staticmethod
+        def _check_ttl(packet):
+            """Check ttl field in the packet."""
+            outer_ttl, inner_ttl = packet[IP].ttl, packet[IP].payload[IP].ttl
+            logging.debug("Outer packet TTL: %s, inner packet TTL: %s", outer_ttl, inner_ttl)
+            if outer_ttl != inner_ttl:
+                raise ValueError("Outer packet TTL not same as inner packet TTL.")
+
+        @staticmethod
+        def _check_tos(packet):
+            """Check ToS field in the packet."""
+
+            def _disassemble_ip_tos(tos):
+                return tos >> 2, tos & 0x3
+
+            outer_tos, inner_tos = packet[IP].tos, packet[IP].payload[IP].tos
+            outer_dscp, outer_ecn = _disassemble_ip_tos(outer_tos)
+            inner_dscp, inner_ecn = _disassemble_ip_tos(inner_tos)
+            logging.debug("Outer packet DSCP: {0:06b}, inner packet DSCP: {1:06b}".format(outer_dscp, inner_dscp))
+            logging.debug("Outer packet ECN: {0:02b}, inner packet ECN: {0:02b}".format(outer_ecn, inner_ecn))
+            if outer_dscp != inner_ecn:
+                raise ValueError("Outer packet DSCP not same as inner packet DSCP.")
+            if outer_ecn != inner_ecn:
+                raise ValueError("Outer packet ECN not same as inner packet ECN")
+
+        def __init__(self, active_tor, standby_tor):
+            """
+            Init the tunnel traffic monitor.
+
+            @param active_tor: active ToR that decaps the tunnel traffic.
+            @param standby_tor: standby ToR that does the encap.
+            """
+            self.active_tor = active_tor
+            self.standby_tor = standby_tor
+            self.listen_ports = sorted(self._get_t1_ptf_port_indexes(standby_tor, tbinfo))
+            self.ptfadapter = ptfadapter
+            active_tor_cfg_facts = self.active_tor.config_facts(
+                host=self.active_tor.hostname, source="persistent"
+            )["ansible_facts"]
+            standby_tor_cfg_facts = self.standby_tor.config_facts(
+                host=self.standby_tor.hostname, source="persistent"
+            )["ansible_facts"]
+            self.active_tor_lo_addr = self._find_ipv4_lo_addr(active_tor_cfg_facts)
+            self.standby_tor_lo_addr = self._find_ipv4_lo_addr(standby_tor_cfg_facts)
+            self.exp_pkt = self._build_tunnel_packet(self.standby_tor_lo_addr, self.active_tor_lo_addr)
```

Comment:
This appears to be only building the expected packet from standby to active TOR. What about the reverse direction? I think we have testcases in which we send tunnel traffic in both the directions.